### PR TITLE
Support for HD Wallet

### LINKED ISSUE #81: Add BIP-39 +  SLIP-10 for recovering wallet
Add the memorable words feature to recover a wallet/key, so users who forget their login details (or switch to a new computer) can regain access. 

We want to implement these two in order to provider both recoverable wallet using a mnemonic, but also hereditary deterministic key generation.

SLIP-10 is pretty much an extension of BIP-32 which enable it to work with ed25519 (https://github.com/satoshilabs/slips/blob/master/slip-0010.md)

There does not seems to be much implementation in golang, but the spec is fairly simple. It's also used by a couple of other wallet, and also hard wallet like Trezor.

This Issue will require to change the format of the wallet, we'll need to decided if we want to make them  retrocompatible too (this is a completely different / new way of generating public key), maybe having such a breaking change now, is not much of an issue until we get to mainnet.

#Acceptance

### PR DESCRIPTION
**Pull request by ValentinTrinque**
Close #81

### CODE REVIEW

File: wallet/hd_wallet.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+		keys[i] = key.DeepCopy()
+	}
+	return keys
+}
+
+// GenerateKeyPair generates a new key pair from a node, that is derived from
+// the wallet node.
+func (w *HDWallet) GenerateKeyPair() (KeyPair, error) {
+	nextIndex := w.keyRing.NextIndex()
+	childNode, err := w.node.Derive(KeyStartIndex + nextIndex)
```
Comment (jeremyletang) [inline wallet/hd_wallet.go:123]: Shouldn't this be OriginIndex+ nextIndex?

Comment (jeremyletang) [reply wallet/hd_wallet.go:123]: If not it means that the first key you generate will be before the one you create when creating the wallet, and you would end up with a collision.
When creating the wallet you use -> OriginIndex which is KeyStartIndex + Magic.
But here you use KeyStartIndex + nextIndex, I imagine at some point nextIndex would be equal to nextIndex, and then you get a collision?

**Review by jeremyletang — APPROVED** [review]
Just one comment which needs to be addressed I think, but once done please merge.

File: wallet/hd_wallet.go (hunk 1 above)
Comment (ValentinTrinque) [reply wallet/hd_wallet.go:123]: Nice catch, you're right ! This is a left over. I thought I removed it...

**Pull request merged**